### PR TITLE
Add maxPods/CPU/MEM SystemReservedResource configuration item

### LINKED ISSUE #1832: Add edged with more personalized options
**What would you like to be added/modified**:
When I use Edged, I want to set the host reserve resource limit (CPU, memory, POD number, etc.).But it doesn't look like there are many configuration items involved (for example, no CPU limit configuration). Can we add more edged personalized configurations like Kubelet in the future?
We can refer to kubelet's configuration item:
kubelet:https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubelet/config/v1beta1/types.go
**Why is this needed**:
With more control over the edged node, configure it as needed

### PR DESCRIPTION
**Pull request by qingchen1203**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Add maxPods/CPU/MEM SystemReservedResource configuration item
**Which issue(s) this PR fixes**:
Fixes #1832

**Special notes for your reviewer**:
Change some API items
**Does this PR introduce a user-facing change?**:
NONE
```release-note

```

### CODE REVIEW

File: edge/pkg/edged/edged_status.go [hunk 1]
```diff
@@ -360,6 +368,20 @@ func (e *edged) setCPUInfo(total, allocated v1.ResourceList) error {
...
 	allocated[v1.ResourceCPU] = total[v1.ResourceCPU].DeepCopy()
 
+	cpu := resource.MustParse(fmt.Sprintf("%d", runtime.NumCPU()))
+	total[v1.ResourceCPU] = cpu.DeepCopy()
+
+	reservedCPU := constants.DefaultSystemReservedCPU
+	if config.Config.SystemReserved["cpu"] != "" {
+		reservedCPU = config.Config.SystemReserved["cpu"]
+	}
+	reservationCPU := resource.MustParse(reservedCPU)
```
Comment (GsssC) [inline edge/pkg/edged/edged_status.go:378]: ditto

File: edge/pkg/edged/edged_status.go [hunk 2]
```diff
@@ -348,6 +351,11 @@ func (e *edged) setMemInfo(total, allocated v1.ResourceList) error {
 	mem := resource.MustParse(strconv.FormatInt(totalMem, 10) + "Mi")
 	total[v1.ResourceMemory] = mem.DeepCopy()
 
+	reservedMem := constants.DefaultSystemReservedMEM
+	if config.Config.SystemReserved["memory"] != "" {
+		reservedMem = config.Config.SystemReserved["memory"]
+	}
+	reservationMemory := resource.MustParse(reservedMem)
```
Comment (GsssC) [inline edge/pkg/edged/edged_status.go:358]: Avoid redundant variables
```suggestion
	var  reservationMemory resource.Quantity
	if config.Config.SystemReserved["memory"] != "" {
		reservationMemory = resource.MustParse(config.Config.SystemReserved["memory"])
	}else{
		reservationMemory = resource.MustParse(constants.DefaultSystemReservedMEM)
	}
```

File: pkg/apis/componentconfig/edgecore/v1alpha1/types.go [hunk 3]
```diff
@@ -177,6 +177,13 @@ type Edged struct {
 	// default false,
 	// Note: Can not use "omitempty" option, it will affect the output of the default configuration file
 	GPUPluginEnabled bool `json:"gpuPluginEnabled"`
+	// MaxPods indicates max num pods that can run on this edged
+	// default 110
+	MaxPods int32 `json:"maxPods,omitempty"`
+	// A set of ResourceName=ResourceQuantity (e.g. cpu=200m,memory=100Mi) pairs
+	// that describe resources reserved for non-kubernetes components.
+	// default cpu=0m,memory=100Mi
+	SystemReserved map[string]string `json:"systemReserved,omitempty"`
```
Comment (GsssC) [inline pkg/apis/componentconfig/edgecore/v1alpha1/types.go:186]: Add validation in `pkg\apis\componentconfig\edgecore\v1alpha1\validation\validation.go`

File: edge/pkg/edged/edged_status.go [hunk 4]
```diff
@@ -348,6 +351,11 @@ func (e *edged) setMemInfo(total, allocated v1.ResourceList) error {
 	mem := resource.MustParse(strconv.FormatInt(totalMem, 10) + "Mi")
```
Comment (GsssC) [inline edge/pkg/edged/edged_status.go:351]: The variables name of this func is awful. Change 

- `total` to `capcility`, 
- `allocated` to `allocatable`, 
- `m` to `machineMem`, 
- `totalMem ` to `machineMemKi`, 
- `mem` to `machineMemMi`

File: edge/pkg/edged/edged_status.go [hunk 5]
```diff
@@ -360,6 +368,20 @@ func (e *edged) setCPUInfo(total, allocated v1.ResourceList) error {
 	total[v1.ResourceCPU] = resource.MustParse(fmt.Sprintf("%d", runtime.NumCPU()))
 	allocated[v1.ResourceCPU] = total[v1.ResourceCPU].DeepCopy()
```
Comment (GsssC) [inline edge/pkg/edged/edged_status.go:369]: change variables name as well

**Comment by GsssC** [comment]
https://github.com/kubernetes/community/blob/master/contributors/design-proposals/node/node-allocatable.md
https://kubernetes.io/zh/docs/tasks/administer-cluster/reserve-compute-resources/
According to above two links, do we need add **kubeReserve** for edgecore?

**Comment by GsssC** [comment]
> System-Reserved (proposed) - Compute resources reserved for processes which are not managed by Kubernetes. Currently this covers all the processes lumped together in the /system raw container.
>Kube-Reserved (proposed) - Compute resources reserved for Kubernetes components such as the docker daemon, kubelet, kube proxy, etc.

Are edgecore now belong to **System-Reserved** or **Kube-Reserved like let**?  : )
@fisherxu

File: edge/pkg/edged/edged_status.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (qingchen1203) [reply edge/pkg/edged/edged_status.go:358]: Thanks for your suggestion, I will fix the redundant variable names

**Comment by fisherxu** [comment]
> Are edgecore now belong to System-Reserved or Kube-Reserved like let? : )
@fisherxu

No, we haven't enabled these options yet... @GsssC  But it should be needed.

**Comment by GsssC** [comment]
@fisherxu In case of **System-Reserved** , are there some codes in let but not in edgecore, which control **cgroup** and ensure process (belong to system) has engouth resource?

**Comment by GsssC** [comment]
@qingchen1203 @fisherxu 
overall lgtm. Besides, `EdgeCoreCgroups` and `SystemCgroups` is set to `""` in default by pkg/apis/componentconfig/edgecore/v1alpha1/types.go`. According to 

https://kubernetes.io/zh/docs/tasks/administer-cluster/reserve-compute-resources/#

> 请注意，如果 --system-reserved-cgroup 不存在，Kubelet 不会创建它。如果指定了无效的 cgroup，Kubelet 将会失败。明确保留的 CPU 列表
>
>请注意，如果 --kube-reserved-cgroup 不存在，Kubelet 将不会创建它。如果指定了一个无效的 cgroup，Kubelet 将会失败。

May they need to be set as kubelet like.

code in `pkg/kubelet/cm/container_manager_linux.go` is twist, need more local test as well. Thanks : )

**Comment by qingchen1203** [comment]
thanks for you comment, i will do more local test for your concern

**Comment by qingchen1203** [comment]
@GsssC thanks for you comment, i will do more local test for your concern, and give the test results later

**Comment by qingchen1203** [comment]
> > System-Reserved (proposed) - Compute resources reserved for processes which are not managed by Kubernetes. Currently this covers all the processes lumped together in the /system raw container.
> > Kube-Reserved (proposed) - Compute resources reserved for Kubernetes components such as the docker daemon, kubelet, kube proxy, etc.
> 
> Are edgecore now belong to **System-Reserved** or **Kube-Reserved like let**? : )
> @fisherxu
When the kubelet parameters kube-reserved and kube-reserved-cgroup are not specified, kublet belongs to system-reserved. 
The figure below shows the cgroup control group where kublet.service is located when only kubelet parameters system-reserved and system-reserved-cgroup are specified.
image

**Comment by qingchen1203** [comment]
@GsssC @fisherxu 
I encountered some confusing phenomena during the test and found that the resource usage of pod still cannot be restricted in kubelet.
Here is my cluster distribution structure and kublet configuration:
image
image

The available resources of the api-server reported by the kubelet node are as follow:
image

Follow this article for systetem-reserved resource limit test:
[How to set Kubernetes resource limits](https://developer.aliyun.com/article/608931)

The following are two deploy configuration files for cpu and mem resource limit tests. For testing, set the resource limit and other parameters as extreme cases:
image
image

Ideally, the resource usage of the pod is:
`cpu-test: `
`cpu-usage = Capacity[cpu] - system-reserved[cpu] = Allocatable[cpu] = 1000m`
`mem-test: `
`mem-usage = Capacity[memory] - system-reserved[memory] = Allocatable[memory] = 224640Ki = 219.3Mi`

But the actual situation of resource usage is different from the ideal situation:
cpu-usage
image
mem-usage
image

The usage of cpu and mem in the test is greater than the ideal value above, and the system-reserved parameter has no effect in kubelet. So I am confused, what value is set by system-reserved ?
From the name point of view, the two parameters are reserved for resource limitation to ensure that the system components or k8s components have enough resources to run stably. But when the kubelet node enters the cgroup, it can be found, the system-reserved[memory] value is set to the value of memory.limit_in_bytes, which means that system-reserved becomes "system-limited".
image

So I have 3 questions:
Q1: Does the parameter system-reserved reserve resources for system components or limit the resource usage of system components?
Q2: Resource restriction scenarios that are not implemented in kubelet do not need to be implemented in edgecore, right?
Q3: If I want to ask questions on this issue in the k8s community, which SIG should I look for?

File: pkg/apis/componentconfig/edgecore/v1alpha1/validation/validation.go [hunk 6]
```diff
@@ -18,6 +18,7 @@ package validation
 
 import (
 	"fmt"
+	"k8s.io/apimachinery/pkg/api/resource"
```
Comment (GsssC) [inline pkg/apis/componentconfig/edgecore/v1alpha1/validation/validation.go:21]: sort imports, mv to the second part.

**Review by GsssC — COMMENTED** [review]
Plz squash commits

**Review by GsssC — COMMENTED** [review]
image
Make sure if we need mkdir for cgroup.

**Comment by GsssC** [comment]
plz resolve confilict
:)

**Comment by qingchen1203** [comment]
> plz resolve confilict
> :)

Conflict resolved

**Comment by kevin-wangzefeng** [comment]
We still want this feature

**Review by GsssC — COMMENTED** [review]
/lgtm
/assign @fisherxu

**Comment by huapox** [comment]
+1

**Pull request closed without merging**